### ISSUE TITLE
Allow RichTextEffects to insert or replace text

### ISSUE THREAD

**Issue by SekoiaTree** [godotengine/godot-proposals#6078]
### Describe the project you are working on

A card game using RichTextLabel for displaying card descriptions.

### Describe the problem or limitation you are having in your project

It'd be convenient to be able to write something like `[stat damage]` in the description of a card to automatically fetch that stat from the card and display it. However, the current `RichTextEffect` API does not allow to change text in any way, only colors and positions.

Additionally, it is currently impossible to do anything analogous to the hardcoded `[lb]` and similar tags, which are part of a large if-else chain in the engine code.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Add a second virtual function to `RichTextEffect`, named something like `_process_custom_replacement` or similar, which has its own analogue to `CharFXTransform`. That analogue, say `CharFXReplace`, would have a string `text` which is the text included within the tag.

Additionally, `RichTextEffect`s should be able to be marked as "solo" tags which don't need a closing tag (maybe another class or so)

Alternatively, tags which end with `/]` could be considered as "solo" tags, similar to HTML (although this is generally not a part of bbcode), and whether the tag is solo could be a parameter to the virtual functions.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

The `lb` tag could be implemented as
```
extends RichTextEffect
class_name RichTextLb

var bbcode = "lb"
var solo = true

def _process_custom_replacement(CharFXReplace replace):
  replace.text = "["
  return true
```

This *would* be less efficient than the current version. I'm open to hearing ideas on how to improve performance for these.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

This can be worked around, essentially by reimplementing bbcode manually. However, this is both ugly and duplicating code from the engine. Additionally, this suggestion could also reduce the very large amount of if-elses in the rich_text_label code (a whopping 700 lines).

I'm not satisfied with my solution as a whole, however. I would like to hear better ideas on this. My goal is to make it possible to reimplement any of the base hardcoded tags in GDScript.

### Is there a reason why this should be core and not an add-on in the asset library?

As core, it would allow reducing the complex and quite messy code of `rich_text_label`. It could be an add-on, but it would be code duplication.

I am willing to code this myself and do have a very basic implementation already in.

**Comment by dalexeev** [comment]
You can use the `meta_hover_started` signal to display the tooltip. You can also implement a BBCodes "preprocessor" to replace your tags with meta (`[url]`) tags, like this:

```gdscript
func replace_bbcodes(text: String) -> String:
    for damage_type in Damage.Type:
        var color: Color = Damage.get_label_color(Damage.Type[damage_type])
        damage_type = damage_type.to_lower()
        text = text.replace(
            '[damage=' + damage_type + ']',
            '[url=card:damage/' + damage_type + '][color=' + color.to_html() + ']',
        )
    text = text.replace('[/damage]', '[/color][/url]')
    
    return text

# [damage=physical]10[/damage] -> [url=card:damage/physical][color=#999999]10[/color][/url]
```

Related:

* #5880

**Comment by SekoiaTree** [comment]
Yep, I mentioned preprocessing/reimplementing bbcode in the "workaround" section. Mostly this is to reduce the very large if-else block in the engine code, and bring a bit more compat between the built-in bbcode and what we can do with custom bbcodes.
